### DISCUSSION TITLE
How to delete users?

### DISCUSSION THREAD

**Discussion by yummyweb** [supabase/supabase#250] [Questions] [answered]
hey guys, how can i delete users from authentication tab in supabase?

**Comment by kiwicopple** [comment, marked as answer]
Hey @antriksh123 - we will add a button for this soon, but at the moment you can run this query:

```sql
delete from auth.users
where id = THE_USERS_UID
```

**Reply by yummyweb** [reply]
Thanks. I'll be using supabase for a really big project. And i just want to tell you that I have made a post on supabase on blog here - https://thewebtech.hashnode.dev/open-source-firebase. Your product is more than firebase. I also saw your DEV posts and youtube video.

**Reply by kiwicopple** [reply]
Wow, that's amazing feedback thanks @antriksh123 !

Also a great blog post. Perhaps you can include the table editor too :) 

![Screenshot 2020-10-14 at 11 53 42 AM](https://user-images.githubusercontent.com/10214025/95941755-0cb11a00-0e14-11eb-84a7-6d097ecccc78.png)

**Reply by yummyweb** [reply]
yeah i just put the image of sql editor. I wanted to ask you that i can run that sql query u gave me but how i do i execute it in javascript in my code. like is there a query() function.

**Reply by kiwicopple** [reply]
Ah, perhaps you can create a stored procedure, then run it as an RPC call?  https://supabase.io/docs/library/stored-procedures

**Reply by yummyweb** [reply]
hey another thing. can u educate me more on managing users because i dont understand how to like get the email and uuid of the user. here is my code
```
const getUser = async () => {
    let { body: user } = await supabase.auth.user()

    return user
  }
```

the user returned gives me the same function as iutput meaning that when i execute 
```
let user = getUser()
console.log(user)
```
I get the function as a string. When i do somehting like:
```
user()
```

then it return a promise. I am confused. Please Help!

**Reply by kiwicopple** [reply]
Can you see if this returns what you're looking for? `console.log('current user', supabase.auth.currentUser)`

**Reply by libovness** [reply]
@kiwicopple Can you think of a reason why `Auth.useUser()` and `supabase.auth.api.getUserByCookie(req)` would return the user but `supabase.auth.currentUser` would be `null`? currently experiencing this, surely my fault or some young Padawan mistake, but for the life of me I can't get requests past my policy (assuming that `auth.uid()` relies on `supabase.auth.currentUser`

**Reply by kiwicopple** [reply]
Within a browser context right?

> Auth.useUser()

Where is this one coming from? Can you share your code?

**Reply by libovness** [reply]
yes this is within the browser context

I *think* I may have found the issue and it's PBCAK. I'm using NextJS. A page is calling my api in the `/api` directory, which contains the Supabase API call. 

Within my page, `supabase.auth.currentUser` contains the user, but within the api route `supabase.auth.currentUser` is `null`. That would seem to make sense as the API route is built separately, though it's pattern that (I think?) would be nice to support if it makes sense?

**Reply by libovness** [reply]
ok now i get that it doesn't make sense to call Supabase from an `api` route using the JS SDK (if I wanted to do that I could just construct the call via `fetch`). i'm an idiot and you're awesome @kiwicopple because supabase is awesome

**Reply by MarkOliv** [reply]
is there any way to delete the user by my application? example: My app has a profile page and there is an option to delete the account

**Reply by MadeByDouglas** [reply]
I also had this question, how is it done?

**Reply by oskarrough** [reply]
@MarkOliv @MadeByDouglas what's not working, what did you try?

If you can delete the user from your backend, you can run the query above.

If however you're on the client, to avoid being able to delete users that are not yourself, 1) create a PostgreSQL function and 2) call it from your client.

```sql
CREATE or replace function delete_user()
	returns void
LANGUAGE SQL SECURITY DEFINER
AS $$
	--delete from public.profiles where id = auth.uid();
	delete from auth.users where id = auth.uid();
$$;
```

```js
await supabaseClient.rpc('delete_user');
```

**Reply by TechBoyy6** [reply]
is there a way to do this with api calls?

**Reply by oskarrough** [reply]
@TechBoyy6 that's what the last line does for you, the `supabaseClient.rpc('my_postgres_function')` part. You can wrapit in another API if you like.

**Reply by ycagri** [reply]
@oskarrough I think you also need to call `supabaseClient.auth.signOut(scope: SignOutScope.global)` after successful deletion. Otherwise, app goes into a instable state like user does not exists but auth says user is logged in.

**Reply by shivpratik** [reply]
Will this work with Anon key?

**Reply by jacobcoccari** [reply]
No, it won't work with the anon key. It needs to be an admin user, which is two levels above. In Supabase, there is anon, authenticated, and admin users.

**Reply by AlienKevin** [reply]
For additional security, you might consider setting `SET search_path = ''` to prevent any potential search path hijacking. Though in this case, I'm not sure if that's possible or not.

**Reply by adt2** [reply]
Realize I'm a little late to the party here, but I'm trying to implement the whole "let the user delete their account" system and I'm having a little trouble. I've added the "delete_user" function above, and I'm calling it from my Flutter front-end via `_supabaseClient.rpc('delete_user');`, but I'm not getting any action. Nothing happens on the database end, and the only that happens on the application end is the logout call (_supabaseClient.auth.signOut(scope: SignOutScope.global).

The only thing I can think of is the `...where id = auth.uid;` statement; where does 'id' come from? I get that `auth.uid` comes from the auth.users table, but I'm not following where the left side `id` is coming from.

**Reply by ycagri** [reply]
@adt2 id comes from `client.auth.currentUser.id` where `client` is `SupabaseClient`. Do you use `id` column of `auth.users` as foreign key in any other table without using `on delete cascade`?

**Reply by adt2** [reply]
@ycagri turns out I *did* have the auth.id referenced as a FK elsewhere without using `on delete cascade`. I've updated that but my "delete yourself" feature still isn't working.

I've got the auth.id field used as a default value in several places as well - would that prevent a row from being deleted? Not FK references, mind you - just default values in an `owner` field in several tables.

**Reply by ycagri** [reply]
@adt2 That should not be a problem. However, it is not a good practice to use `id` of `auth.users` as default value without using foreign key constraint. You probably will end up having orphan records after deletion. Just check Postgres logs, it seems like you have some sort of integrity constraint problem.

**Reply by adt2** [reply]
@ycagri here's a dumb question...should I be passing the user's id as a parameter when calling the `delete_user` function?

**Reply by ycagri** [reply]
@adt2 not necessarily. you can define a function as below if you want to allow user to delete only his/her account:
```
create or replace function delete_user()
returns void as
$$
begin
delete from auth.users where id = (select auth.uid());
end;
$$ language plpgsql security definer set search_path = '';
```

**Reply by adt2** [reply]
I don't see *any* errors in any of the logs, and no mention of the `delete_user` function being called (although the `logout` call which directly follows it *is* being called and logged). Seems like I have a configuration issue somewhere.

**Reply by adt2** [reply]
@ycagri That did it! Thanks! Your code is slightly different than the original function as entered way up above. Yours says `...where id = (select auth.uid());`; the one above leaves out the `select` part. Changing the code fixed my problem. 

Thanks again!

**Reply by darkblitzrc** [reply]
> @MarkOliv @MadeByDouglas what's not working, what did you try?
> 
> If you can delete the user from your backend, you can run the query above.
> 
> If however you're on the client, to avoid being able to delete users that are not yourself, 1) create a PostgreSQL function and 2) call it from your client.
> 
> ```sql
> CREATE or replace function delete_user()
> 	returns void
> LANGUAGE SQL SECURITY DEFINER
> AS $$
> 	--delete from public.profiles where id = auth.uid();
> 	delete from auth.users where id = auth.uid();
> $$;
> ```
> 
> ```js
> await supabaseClient.rpc('delete_user');
> ```

thank you so much man. i LOVE YOU

**Reply by MadeByDouglas** [reply]
yes I got it to work using that approach, once I understood how the pieces fit together (postgres, client etc.)

**Comment by MadeByDouglas** [comment]
@oskarrough  Sorry for the delay; well I'm not sure what to try, thats part of the problem, I just don't understand enough of how these pieces fit together. Is there like a Postgres for beginners course or something really basic that Supabase can provide to front end or mobile devs like myself?

What I think you're saying is, I can write that Postgres query on my Supabase admin panel and then call it from my Swift client to trigger it which will delete the account based on matching the user id from the swift client to whats found on Supabase? So then I just write that as a Postgres query, no edge function needed?

**Reply by oskarrough** [reply]
There are a lot of parts to understand, for sure.

But yes. Define a Postgres function (via SQL) and call it with `supabase.rpc` from your app. If you're using swift, see https://supabase.com/docs/reference/swift/rpc

**Reply by MadeByDouglas** [reply]
Ah ok, this is super helpful! I think I have it sorted now, simpler than I thought now that you've explained it, thanks so much. That documentation link is great as well, thank you.

**Comment by k-thornton** [comment]
just fyi to anyone stumbling upon this while working with the API:
https://supabase.com/docs/reference/javascript/auth-admin-deleteuser

**Comment by anthonycmain** [comment]
I wrote this a few weeks ago if you stumble across this thread and are looking for a soft delete approach instead

https://thedistance.co.uk/insights/how-to-handle-soft-deletes-in-supabase-using-sql
